src/main/java/com/julienviet/pgclient/Numeric.java

```diff
@@ -0,0 +1,162 @@
+   */
+  public static Numeric create(Number number) {
+    if (number == null) {
+      throw new NullPointerException();
+    }
+    if (number instanceof Double && ((Double)number).isInfinite() || number instanceof Float && ((Float)number).isInfinite()) {
+      throw new NumberFormatException("Infinite numbers are not valid numerics");
+    }
+    return new Numeric(number);
+  }
+
+  /**
+   * Parse and return a {@code Numeric} instance for the given {@code s}.
+   * <p/>
+   * The string {@code "Nan"} will return the {@link #NaN} instance.
+   *
+   * @param s the string
+   * @return the {@code Numeric} value
+   */
+  public static Numeric parse(String s) {
+    switch (s) {
+      case "NaN":
+        return NaN;
+      default:
+        return new Numeric(new BigDecimal(s));
+    }
+  }
+
+  private Numeric(Number value) {
+    this.value = value;
+  }
+
+  @Override
+  public int intValue() {
+    return value.intValue();
+  }
+
+  @Override
+  public long longValue() {
+    return value.longValue();
+  }
+
+  @Override
+  public float floatValue() {
+    return value.floatValue();
+  }
+
+  @Override
+  public double doubleValue() {
+    return value.doubleValue();
+  }
+
+  /**
+   * @return {@code true} when this number represents {@code NaN}
+   */
+  public boolean isNaN() {
+    return value instanceof Double && ((Double)value).isNaN() || value instanceof Float && ((Float)value).isNaN();
+  }
+
+  /**
+   * @return  the numeric value represented by this object after conversion
+   *          to type {@code BigDecimal}.
+   */
+  public BigDecimal bigDecimalValue() {
+    if (value instanceof BigDecimal) {
+      return (BigDecimal) value;
+    } else if (value instanceof BigInteger) {
+      return new BigDecimal((BigInteger)value);
+    } else if (isNaN()) {
+      return null;
+    } else {
+      return new BigDecimal(value.toString());
+    }
+  }
+
+  /**
+   * @return  the numeric value represented by this object after conversion
+   *          to type {@code BigInteger}.
+   */
+  public BigInteger bigIntegerValue() {
+    if (value instanceof BigInteger) {
+      return (BigInteger) value;
+    } else if (value instanceof BigDecimal) {
+      return ((BigDecimal)value).toBigInteger();
+    } else if (isNaN()) {
+      return null;
+    } else {
+      return new BigInteger(Long.toString(value.longValue()));
+    }
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj instanceof Numeric) {
+      Numeric that = (Numeric) obj;
+      if (value.getClass() == that.value.getClass()) {
+        return value.equals(that.value);
+      } else {
+        BigDecimal l = bigDecimalValue();
+        BigDecimal r = that.bigDecimalValue();
+        if (l == null) {
+          return r == null;
+        } else if (r == null) {
+          return false;
+        }
+        return l.compareTo(r) == 0;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    return intValue();
```

Comment:
Interesting to base it on the int value. It'll cause a poor distribution though if you got multiple numerics (1.1, 1.2 etc.) in a small range, perhaps rather base it on the `BigDecimal` value of this (it might also be calculated once and cached)?

Reply:
I've done that because it provides a consistent value between Number

Reply:
to have `Number.create(4).hashcode() == Number.create(new BigInteger(4)).hashcode()`

Reply:
I see. Is there a need for that, though? These values aren't considered as equal as per `equals()`, so it seems advantageous to put them into different hash buckets actually? I was just wondering about a case where you'd end up with lots of closely related decimal values in the same hash bucket. Not sure whether it makes a difference in practice :)

Reply:
I think we need it because equality implies hashcode are the same

Reply:
Ah yes, indeed you _do_ consider the two Numerics as equal, so they must have the same hash. You'd have a better value distribution by basing the hash on the BigDecimal value, but the current impl should work.